lib/resources/forms.js

```diff
@@ -136,6 +136,18 @@ module.exports = (service, endpoint) => {
       .then(((form) => Promise.all([ Forms.publish(form), Submissions.clearDraftSubmissions(form.id) ])))
       .then(success)));
 
+  // Entity/Dataset-specific endpoint that is used to show how publishing
+  // a form will change any datasets mentioned in the form.
+  // Even though there are dataset-related permissions, we will use the form modification verbs
+  // for authentication because it seems weird to mix the ability/inability to alter a dataset
+  // with the ability to modify forms -- they should just be the same.
+  service.get('/projects/:projectId/forms/:id/draft/dataset-diff', endpoint(({ Forms, Datasets }, { params, auth }) =>
+    Forms.getByProjectAndXmlFormId(params.projectId, params.id, false, Form.DraftVersion)
+      .then(getOrNotFound)
+      .then((form) => auth.canOrReject('form.update', form)) // use form verb
```

Comment:
This would be something like this:

```js
.then((form) => Promise.all([
  auth.canOrReject('form.read', form)),
  auth.canOrReject('dataset.list', project))
]))
```

You'd need to get the project first.

Reply:
I realized I can use the form here without another query to look up the project and the auth will implicitly look up the actee parent chain to the form's parent, which is the project or higher.

Reply:
Oh interesting! I can see the appeal of avoiding an extra query, though making an optimization around auth makes me a little nervous. I think this approach would work well with typical Central use cases (and everything in Frontend), but wouldn't work as well with behavior that we expose over the API. In the API docs, we advertise the existence of a:

> Form-specific Assignments resource at `projects/…/forms/…/assignments`, allowing granular role assignments on a per-Form basis.

For example, Frontend only lets you create `viewer` users on a project level, but I believe the API allows you to create a `viewer` to a single form. But now that the `viewer` role grants `dataset.list`, checking for `dataset.list` on the form rather than the project would provide unintentional access to information about datasets:

- Alice is an admin and makes Chelsea a `viewer` on a form in a project that has a dataset
- Chelsea shouldn't be able to access &hellip;/dataset-diff for the form, because she doesn't have any project-level access, so doesn't have access to the dataset.
- However, checking for `dataset.list` on the form will allow her to access the endpoint, because the `viewer` role grants her that verb.

Reply:
good point! did some TDD to add this test case, saw it in action, and changed the code.